transition `@timeit` in `Core.Compiler` to use tracy instead

KristofferC opened the pull request:
This allows us to see the already instrumented compiler passes in tracy, cf:

<img width="680" alt="image" src="https://user-images.githubusercontent.com/1282691/236695695-576a86ce-c461-42db-a576-51ea79d857a1.png">

pchintalapudi commented:
I would like to see this be extensible to ITTAPI/VTune integration as well as Tracy, do you have thoughts on how that would work here? ITTAPI requires the ability to store a field/pointer as timer state.

KristofferC commented:
> ITTAPI requires the ability to store a field/pointer as timer state.

Could you elaborate on what that means more concretely?

KristofferC commented:
I changed it so it should be straight forward (I think) to hook up ITTAPI. Please have a look @pchintalapudi

pchintalapudi commented:
I feel like it would be better to have the zone macro ccall a pair of functions in timing.c with the relevant data that Tracy uses, and then hook in the relevant profiling calls in that pair of functions. I think it would be possible to completely reuse most of the already available jl_timing_block_t code for the start and end, which gets us equal integration with C events for free, and keeps the jl_timing_events/jl_timing_owners as the definitive list of events that we track. We'll probably need a clean way of translating the event from Julia to C, but that could be as simple as binary searching on the event name.

KristofferC commented:
I don't see many advantages to that. The majority of the new code is "scaffolding" (rooting, types etc) which is needed even if the call would be into the existing C-timing infrastructure. The actual ccall stuff is quite minimal. And having to go into C and add categories vs just slapping a `@zone "name" begin` in the Julia code makes it more awkward to use. But if you have some example implementation of it, I'd be happy to look at it.

pchintalapudi commented:
@KristofferC I have an example one here (https://github.com/JuliaLang/julia/commit/c627096e4c8ab71718a10bf08b4d2338fee8bfc6 , branched off #49679 ), with WITH_TIMING_COUNTS output below. Not too sure on the tracy details, but since it mostly reuses the jl_timing_block_t implementation I think it should just work. 

![image](https://user-images.githubusercontent.com/34727397/236961075-9cf2fca5-0c3f-4233-bc2c-48bdd7fb262f.png)

KristofferC commented:
@pchintalapudi, I think the strategy in that commit is a slightly flawed. The indices to the arrays are computed at macro expansion time and cached in the compiled code. But these indices are invalid in a new julia process (where the dynamical allocations no longer exists). So you would need to run something at `__init__` to reallocate all these.

I took the liberty to push a commit to that branch that changes the strategy a bit by moving some of the C handling of the Tracy source locations back into Julia (https://github.com/JuliaLang/julia/commit/8117bd759d947ba737ef8072c90d75815f6d7c91) and now it seems to work again.

<img width="959" alt="image" src="https://github.com/JuliaLang/julia/assets/1282691/39726e6b-4483-4563-bb4a-97ee9a4c7870">

KristofferC commented:
@pchintalapudi, I think there is a small problem in your suggestion in that the lookup of zone happens at compile time (macro expansion time) of the function and will persist into new processes where this is no longer valid. In https://github.com/JuliaLang/julia/commit/a7e1ec0dc4519f9f196e4b7f13bae7e538872319 I moved it to happen at runtime, the first time that zone is run. It seems to work with that.

![gnome-shell-screenshot-6y4s8g](https://github.com/JuliaLang/julia/assets/1282691/754e562f-a647-4919-b625-669127433f45)

KristofferC commented:
Based on the API in https://github.com/JuliaLang/julia/commit/c627096e4c8ab71718a10bf08b4d2338fee8bfc6 me and @topolarity looked into instrumenting the actual LLVM passes and came up with https://github.com/JuliaLang/julia/commit/794673b4b4fd517b5bb70c689afc881d2ed8ff89. With that you can get a list like:

![gnome-shell-screenshot-97c87l](https://github.com/JuliaLang/julia/assets/1282691/9cb86e24-d1f1-42a1-bda7-8752a27cf439)

and the zones as

![gnome-shell-screenshot-b7wxmz](https://github.com/JuliaLang/julia/assets/1282691/46e9ebe0-4a2f-4d77-a345-670cfedbbb7c)

We discussed that it is probably better to have the zones themselves be named after the pass which would require getting rid of the hardcoded list of zone names.

pchintalapudi commented:
That's great! I was also thinking about something like that. I would suggest stripping the PassManager<X> passes from the events, since they're just organizational tools and it's denoted in the pass pipeline anyways. I think it would also be useful to use the analysis callbacks as well, to see when particular analysis computations are bad.

KristofferC commented:
>  I would suggest stripping the PassManager passes from the events

How would I identify one of those? Just from the name of them?

KristofferC commented:
Okay, updated with showing the pass name (only in Tracy for now) and also remove the PassManager + PassAdaptor passes.

![gnome-shell-screenshot-clf178](https://github.com/JuliaLang/julia/assets/1282691/2e31b27a-6b7f-4c34-9638-eba340029bc6)

pchintalapudi commented:
WRT compiler, the other thing I want to do is to add pre/post optimization counters for lines of IR (both Julia and LLVM). I think if combined with the timing counts printout we could track how changes to optimization affect amount of code generated. But that requires #49535 to be merged.

KristofferC committed:
wip on tracy in Compiler

KristofferC committed:
wip

KristofferC commented:
Updated

topolarity commented on Compiler/src/profiling/tracy.jl:
```diff
@@ -0,0 +1,67 @@
+    file_sym::Symbol
+end
+TracySrcLoc(zone_name::String, function_name::Symbol, file::Symbol, line::UInt32, color::UInt32) =
+    TracySrcLoc(_strpointer(zone_name),
+                unsafe_convert(Ptr{UInt8}, function_name),
+                unsafe_convert(Ptr{UInt8}, file),
+                line, color, zone_name, function_name, file)
```
```suggestion
    TracySrcLoc(C_NULL, C_NULL, C_NULL, line, color, zone_name, function_name, file)
```

Optional, but simpler since this will be re-initialized anyway

KristofferC committed:
updates

topolarity commented on Compiler/src/profiling/tracy.jl:
```diff
@@ -0,0 +1,67 @@
+        reinit!(loc)
+    end
+    ptr = Ptr{TracySrcLoc}(pointer_from_objref(loc))
+    return ccall((:___tracy_emit_zone_begin, "libTracyClient"), TracyZoneCtx, (Ptr{TracySrcLoc}, Cint), ptr, active)
```
Fair enough - worth a comment though in case anyone removes the interning

topolarity replied:
Ah you're way ahead of me 🙂

topolarity commented on src/timing.h:
```diff
@@ -55,6 +55,10 @@ JL_DLLEXPORT jl_timing_event_t *_jl_timing_event_create(const char *subsystem, c
 JL_DLLEXPORT void _jl_timing_block_init(char *buf, size_t size, jl_timing_event_t *event);
 JL_DLLEXPORT void _jl_timing_block_start(jl_timing_block_t *cur_block);
 JL_DLLEXPORT void _jl_timing_block_end(jl_timing_block_t *cur_block);
+JL_DLLEXPORT int _jl_tracy_enabled(void);
+JL_DLLEXPORT int _jl_ittapi_enabled(void);
+JL_DLLEXPORT int _jl_nvtx_enabled(void);
```
Don't forget to add these to `jl_exported_funcs.inc`

topolarity replied:
Actually, scratch that. DLL-exported, but not re-exported to `libjulia` is probably the right config

topolarity commented on src/timing.c:
```diff
@@ -358,6 +358,31 @@ JL_DLLEXPORT jl_timing_event_t *_jl_timing_event_create(const char *subsystem, c
     return event;
 }
 
+JL_DLLEXPORT int _jl_tracy_enabled(void) {
```
These need to dodge the `#ifdef ENABLE_TIMINGS`

KristofferC committed:
move some defs

KristofferC commented on Compiler/src/profiling/tracy.jl:
```diff
@@ -0,0 +1,67 @@
+
+const srclocs = Vector{TracySrcLoc}()
+
+function _tracy_zone_create(name::String, ex::Expr, linfo::LineNumberNode)
```
How do you suggest we do that?

gbaraldi replied:
I think a lock around it should be fine

gbaraldi replied:
Doesn't look like it :(. Maybe put a isdefined base around it. Kind of ugly but building the compiler is ST only

gbaraldi replied:
@vtjnash any suggestions?

KristofferC replied:
Since the response to this review comment got more or less ignored, I think it is ok to merge it without the locking since:

- The Tracy.jl package does the same
- This is only used in one "package" (Compiler)
- Locks are a bit messy to add at this point
- I don't think lowering runs concurrently at this point?

I don't want this to get stalled again, preventing us from adding more zones and investigating the load time regressions.

gbaraldi replied:
It could be concurrent IIRC. But i think we can just add a profiling lock in C and take it.

gbaraldi commented:
Maybe we make those optional? With an envar or a global?

KristofferC commented:
We already have the `DISABLE_FREQUENT_EVENTS`. We could tap into that.

Commit:
no underscore?

KristofferC committed:
put into modules again

KristofferC committed:
pdate pass names

gbaraldi committed:
Refactor timing locks and expose it to julia

topolarity commented on src/init.c:
```diff
@@ -750,6 +750,7 @@ static void init_global_mutexes(void) {
     JL_MUTEX_INIT(&global_roots_lock, "global_roots_lock");
     JL_MUTEX_INIT(&typecache_lock, "typecache_lock");
     JL_MUTEX_INIT(&profile_show_peek_cond_lock, "profile_show_peek_cond_lock");
+    jl_mutex_timing_lock_init();
```
This seems to initialize this mutex strictly later than before (when it was in `jl_init_timing`)

Any reason why?

topolarity replied:
I think it should be initialized with timing, not later - otherwise an arbitrary subset of timing won't be safe to use during the interim

topolarity commented on Compiler/src/profiling.jl:
```diff
@@ -0,0 +1,38 @@
+
+if WITH_TRACY || WITH_ITTAPI
+    macro zone(name, ex::Expr)
+        lock_timing()
```
This lock should probably be taken just in the implementation of `tracy_zone_create`

gbaraldi committed:
Add suggestions from code review

topolarity commented on Compiler/src/profiling/tracy.jl:
```diff
@@ -0,0 +1,67 @@
+    TracySrcLoc(C_NULL, C_NULL, C_NULL, line, color, zone_name, function_name, file)
+
+@noinline function reinit!(srcloc::TracySrcLoc)
+    lock_timing()
```
I don't think you need the lock here - it's OK for the threads to race since the outcome is idempotent.

It's also insufficient to prevent UB, since you have an unguarded read in begin_event. I'd recommend `@atomic` set/get on these fields instead, which should fix both issues

KristofferC committed:
fix pass names

gbaraldi committed:
Use atomics

KristofferC commented:
`UndefVarError(var=:var"@atomic"` @gbaraldi

KristofferC committed:
Revert "Use atomics"

This reverts commit cf0238edd336ff4167d22d3326f3cad7d97abf06.

KristofferC committed:
Revert "Add suggestions from code review"

This reverts commit 8992ef3635561647337369c2c0085bb5309e2d78.

KristofferC committed:
Revert "Refactor timing locks and expose it to julia"

This reverts commit 15ebde8464c1412fe0fc035f4322b4c60c7eb692.

vtjnash commented on Compiler/src/profiling/tracy.jl:
```diff
@@ -0,0 +1,63 @@
+end
+
+function tracy_zone_begin(loc, active)
+    if loc.zone_name === Ptr{UInt8}(0)
+        reinit!(loc)
+    end
+    # `loc` is rooted in the global `srclocs`
```
This looks like a substantial data race

KristofferC committed:
fix name

topolarity committed:
Add `@atomic` usage again, to avoid data race

This was mostly correct before, but Julia uses `:monotonic` not
`:relaxed`.

This does not resolve the data race in the `push!(...)` to the
`srclocs` Vector, but it's a step in the right direction.

gbaraldi commented on Compiler/src/profiling/tracy.jl:
```diff
@@ -7,9 +7,9 @@ import ..@noinline, ..Cint, ..Vector, ..push!, ..unsafe_convert, ..esc,
-    zone_name::Ptr{UInt8}
-    function_name::Ptr{UInt8}
-    file::Ptr{UInt8}
+    @atomic zone_name::Ptr{UInt8}
```
@atomic doesn't seem to be defined at this point so you will have to do the non macro builting

topolarity replied:
Weird. It built just fine for me locally.

KristofferC replied:
It was `@atomic :symbol` that didn't work I think?

KristofferC replied:
I think it was because this is in a submode but earlier the `@atomc` wasn't imported.

topolarity committed:
import `nothing`

giordano commented:
Would it be possible to document this in the devdocs?  Sounds like a useful feature for developers, but it's completely obscure without diving into the source code.
